From 536378a084c6a4148141e132efee2fa9a464e007 Mon Sep 17 00:00:00 2001
From: Peter Geis <pgwipeout@gmail.com>
Date: Thu, 3 Jun 2021 11:36:35 -0400
Subject: [PATCH] irqchip: gic-v3: add hackaround for rk3568 its

---
 drivers/irqchip/irq-gic-v3-its.c | 70 +++++++++++++++++++++++++++++---
 1 file changed, 65 insertions(+), 5 deletions(-)

--- a/drivers/irqchip/irq-gic-common.h
+++ b/drivers/irqchip/irq-gic-common.h
@@ -34,5 +34,6 @@ extern const struct msi_parent_ops gic_v
 #define RDIST_FLAGS_PROPBASE_NEEDS_FLUSHING    (1 << 0)
 #define RDIST_FLAGS_RD_TABLES_PREALLOCATED     (1 << 1)
 #define RDIST_FLAGS_FORCE_NON_SHAREABLE        (1 << 2)
+#define RDIST_FLAGS_FORCE_NO_LOCAL_CACHE       (1 << 3)
 
 #endif /* _IRQ_GIC_COMMON_H */
--- a/drivers/irqchip/irq-gic-v3-its.c
+++ b/drivers/irqchip/irq-gic-v3-its.c
@@ -2267,6 +2267,11 @@
 {
 	struct page *prop_page;
 
+	if (gic_rdists->flags & RDIST_FLAGS_FORCE_NO_LOCAL_CACHE) {
+		pr_debug("ITS ALLOCATE PROP WORKAROUND\n");
+		gfp_flags |= GFP_DMA;
+	}
+
 	prop_page = its_alloc_pages(gfp_flags,
 				    get_order(LPI_PROPBASE_SZ));
 	if (!prop_page)
@@ -2389,6 +2394,7 @@
 	u64 baser_phys, tmp;
 	u32 alloc_pages, psz;
 	struct page *page;
+	gfp_t gfp_flags;
 	void *base;
 
 	psz = baser->psz;
@@ -2400,8 +2406,10 @@
 		alloc_pages = GITS_BASER_PAGES_MAX;
 		order = get_order(GITS_BASER_PAGES_MAX * psz);
 	}
-
-	page = its_alloc_pages_node(its->numa_node, GFP_KERNEL | __GFP_ZERO, order);
+	gfp_flags = GFP_KERNEL | __GFP_ZERO;
+	if (gic_rdists->flags & RDIST_FLAGS_FORCE_NO_LOCAL_CACHE)
+		gfp_flags |= GFP_DMA;
+	page = its_alloc_pages_node(its->numa_node, gfp_flags, order);
 	if (!page)
 		return -ENOMEM;
 
@@ -3040,6 +3048,10 @@
 {
 	struct page *pend_page;
 
+	if (gic_rdists->flags & RDIST_FLAGS_FORCE_NO_LOCAL_CACHE) {
+		gfp_flags |= GFP_DMA;
+	}
+
 	pend_page = its_alloc_pages(gfp_flags | __GFP_ZERO, get_order(LPI_PENDBASE_SZ));
 	if (!pend_page)
 		return NULL;
@@ -3387,7 +3399,12 @@
 
 	/* Allocate memory for 2nd level table */
 	if (!table[idx]) {
-		page = its_alloc_pages_node(its->numa_node, GFP_KERNEL | __GFP_ZERO,
+		gfp_t gfp_flags = GFP_KERNEL | __GFP_ZERO;
+        	if (gic_rdists->flags & RDIST_FLAGS_FORCE_NO_LOCAL_CACHE) {
+        		gfp_flags |= GFP_DMA;
+        	}
+
+		page = its_alloc_pages_node(its->numa_node, gfp_flags,
 					    get_order(baser->psz));
 		if (!page)
 			return false;
@@ -3471,6 +3488,7 @@
 	unsigned long *lpi_map = NULL;
 	unsigned long flags;
 	u16 *col_map = NULL;
+	gfp_t gfp_flags;
 	void *itt;
 	int lpi_base;
 	int nr_lpis;
@@ -3493,7 +3511,11 @@
 
 	itt = itt_alloc_pool(its->numa_node, sz);
 
-	dev = kzalloc(sizeof(*dev), GFP_KERNEL);
+	gfp_flags = GFP_KERNEL;
+	if (gic_rdists->flags & RDIST_FLAGS_FORCE_NO_LOCAL_CACHE)
+		gfp_flags |= GFP_DMA;
+
+	dev = kzalloc(sizeof(*dev), gfp_flags);
 
 	if (alloc_lpis) {
 		lpi_map = its_lpi_alloc(nvecs, &lpi_base, &nr_lpis);
@@ -4856,6 +4878,21 @@
 	its->vlpi_redist_offset = SZ_128K;
 	return true;
 }
+
+static bool __maybe_unused its_enable_rk3568001(void *data)
+{
+	struct its_node *its = data;
+
+	if (!of_machine_is_compatible("rockchip,rk3566") &&
+	    !of_machine_is_compatible("rockchip,rk3568"))
+		return false;
+
+	its->flags |= ITS_FLAGS_FORCE_NON_SHAREABLE;
+	gic_rdists->flags |= RDIST_FLAGS_FORCE_NON_SHAREABLE |
+			     RDIST_FLAGS_FORCE_NO_LOCAL_CACHE;
+
+	return true;
+}
 
 static bool __maybe_unused its_enable_rk3588001(void *data)
 {
@@ -4943,6 +4980,12 @@
 #endif
 #ifdef CONFIG_ROCKCHIP_ERRATUM_3588001
 	{
+		.desc   = "ITS: Rockchip erratum RK3568001",
+		.iidr   = 0x0201743b,
+		.mask   = 0xffffffff,
+		.init   = its_enable_rk3568001,
+	},
+	{
 		.desc   = "ITS: Rockchip erratum RK3588001",
 		.iidr   = 0x0201743b,
 		.mask   = 0xffffffff,
@@ -5211,6 +5254,7 @@
 {
 	u64 baser, tmp;
 	struct page *page;
+	gfp_t gfp_flags;
 	u32 ctlr;
 	int err;
 
@@ -5245,9 +5289,9 @@
 				&its->phys_base, its->mpidr, svpet);
 		}
 	}
-
+	gfp_flags = GFP_KERNEL | __GFP_ZERO | GFP_DMA;
 	page = its_alloc_pages_node(its->numa_node,
-				    GFP_KERNEL | __GFP_ZERO,
+				    gfp_flags,
 				    get_order(ITS_CMD_QUEUE_SZ));
 	if (!page) {
 		err = -ENOMEM;
